### PR TITLE
Basic Implement of OneDrive Storager

### PR DESCRIPTION
**Pull request by zxy-lgtm**
Sorry for the late pull request.
The past few days I study the onedrive documents and beyondstorage specs, so it takes time to finish a basic one drive client and  some adjust for the latest behavior.
For now, I have implement the basic storager for one drive, but the newest GSP-751 and one-drive conflict with the design of the empty file, the latter doesn't allow empty file exist, I didn’t found ways to avoid that, so I think this could  to be discuss later.
ref: https://stackoverflow.com/questions/64772346/uploading-empty-0-byte-to-sharepoint-online (since the ms voice has shut down …)

TL;DR:
This storage has pass the integration-test except the part of zero size file.

### CODE REVIEW

**Comment by zxy-lgtm** [comment]
2021-09-27 22-39-21 的屏幕截图
2021-09-27 22-38-28 的屏幕截图

**Comment by xxchan** [comment]
Great work! Give me some time to review the details..

**Comment by xxchan** [comment]
I think we can create an error code like below and returns it when size=0, but this doesn't pass integration test. 
```go
var ErrWriteFileEmpty = services.NewErrorCode("onedrive doesn't support upload empty file")
```

I think GSP-751 just considered the API flavor, but doesn't consider the case where writing empty file is not supported... Ping @abyss-w and @JinnyYi

File: Makefile.env [hunk 1]
```diff
@@ -0,0 +1,2 @@
+export STORAGE_ONEDRIVE_INTEGRATION_TEST=on
+export STORAGE_ONEDRIVE_CREDENTIAL=file:<abs_path_of_credential>
```
Comment (xxchan) [inline Makefile.env]: Remove `Makefile.env` from the repo, since in the README it says using `cp Makefile.env.example Makefile.env` to create this file. And add `-include Makefile.env` to `Makefile` so that this file can be used.

File: utils.go [hunk 2]
```diff
@@ -14,14 +29,153 @@ type Storage struct {
...
+	case credential.ProtocolBase64:
+		token, err = base64.StdEncoding.DecodeString(cp.Base64())
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, services.PairUnsupportedError{Pair: ps.WithCredential(opt.Credential)}
+	}
+
+	// create new minio client
```
Comment (xxchan) [inline utils.go]: minio -> onedrive 😄  

BTW I think this comment is redundant
```suggestion
```

File: utils.go [hunk 3]
```diff
@@ -14,14 +29,153 @@ type Storage struct {
...
+
+func (s *Storage) formatObject(v *Item, dir, rp string) *typ.Object {
+	if v.Folder != nil {
+		return s.formatFolderObject(v, dir, rp)
+	}
+
+	return s.formatFileObject(v, dir, rp)
+}
+
+// formatFolderObject format a minio folder object into go-storage object
```
Comment (xxchan) [inline utils.go]: minio -> onedrive

File: utils.go [hunk 4]
```diff
@@ -14,14 +29,153 @@ type Storage struct {
...
+
+	o.ID = filepath.Join(rp, folderName)
+	o.Path = path.Join(dir, folderName)
+
+	o.Mode |= typ.ModeDir
+
+	return o
+}
+
+// formatFolderObject format a minio file object into go-storage object
```
Comment (xxchan) [inline utils.go]: ditto

File: utils_drives.go [hunk 5]
```diff
@@ -0,0 +1,336 @@
+package onedrive
+
+// the refactor of github.com/goh-chunlin/go-onedrive/onedrive upper layer functions(except upload)
```
Comment (xxchan) [inline utils_drives.go]: Can you add some more comments here explaining why the refactoring is needed?

File: utils_drives.go [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+	}
+
+	// get onedrive upload url
+	uploadUrl, err := o.createUploadSession(ctx, absPath, fileSize, description)
+	if err != nil {
+		return
+	}
+
+	// debug log
+	// fmt.Println("create upload session success!")
```
Comment (xxchan) [inline utils_drives.go]: This should not be pushed. (Usually commented code should not be pushed.)

File: utils_drives.go [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+		req, err := http.NewRequestWithContext(ctx, "PUT", uploadUrl, bn)
+		if err != nil {
+			return n, err
+		}
+
+		req.Header.Add("Content-Length", strconv.Itoa(nowSize))
+		req.Header.Add("Content-Range", fmt.Sprintf("bytes %d-%d/%d", n, n+int64(nowSize)-1, fileSize))
+
+		// debug log
+		//fmt.Println(fmt.Sprintf("Content-Length: %s, Content-Range: %s", strconv.Itoa(nowSize), fmt.Sprintf("bytes %d-%d/%d", n, n+int64(nowSize)-1, fileSize)))
```
Comment (xxchan) [inline utils_drives.go]: ditto

File: utils_drives.go [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+		if res.StatusCode != 202 && res.StatusCode != 201 && res.StatusCode != 200 {
+			errContent, _ := ioutil.ReadAll(res.Body)
+			err = fmt.Errorf("upload resource failed, cause %s:%s", res.Status, string(errContent))
+			return n, err
+		}
+		// close body
+		res.Body.Close()
+
+		// debug log
+		//log.Println(fmt.Sprintf("%d of %d done", n+int64(nowSize)-1, fileSize))
```
Comment (xxchan) [inline utils_drives.go]: ditto

**Comment by xxchan** [comment]
The build CI fails, and you can see it's format check: https://github.com/beyondstorage/go-service-onedrive/runs/3722658754 

You can run `make build` before pushing your code to ensure it's formatted :)

**Comment by JinnyYi** [comment]
> I think we can create an error code like below and returns it when size=0, but this doesn't pass integration test.
> 
> ```go
> var ErrWriteFileEmpty = services.NewErrorCode("onedrive doesn't support upload empty file")
> ```
> 
> I think GSP-751 just considered the API flavor, but doesn't consider the case where writing empty file is not supported... Ping @abyss-w and @JinnyYi

We need to specify the expected behavior when writing an empty file is not supported.

For integration test, how about splitting integration test cases like what we did for `Move/Copy`?

**Comment by Xuanwo** [comment]
I think we need to come up with a solution like feature flags or something. Can you just pin the go-integration-test to the older version and leave a comment on this comment? We expect to solve this problem in October.

File: go.sum [hunk 9]
```diff
@@ -1,7 +1,57 @@
...
+cloud.google.com/go/storage v1.5.0/go.mod h1:tpKbwo567HUNpVclU5sGELwQWBDZ8gh0ZeosJ0Rtdos=
+cloud.google.com/go/storage v1.6.0/go.mod h1:N7U0C8pVQ/+NIKOBQyamJIeKQKkZ+mxpohlUTyfDhBk=
+cloud.google.com/go/storage v1.8.0/go.mod h1:Wv1Oy7z6Yz3DshWRJFhqM/UCfaWIRTdp0RXyy7KQOVs=
+cloud.google.com/go/storage v1.10.0/go.mod h1:FLPqc6j+Ki4BU591ie1oL6qBQGu2Bl/tZ9ullr3+Kg0=
+dmitri.shuralyov.com/gpu/mtl v0.0.0-20190408044501-666a987793e9/go.mod h1:H6x//7gZCb22OMCxBHrMx7a5I7Hp++hsVxbQ4BYO7hU=
+github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
+github.com/BurntSushi/xgb v0.0.0-20160522181843-27f122750802/go.mod h1:IVnqGOEym/WlBOVXweHU+Q+/VP0lqqI8lqeDx9IjBqo=
 github.com/Xuanwo/templateutils v0.1.0 h1:WpkWOqQtIQ2vAIpJLa727DdN8WtxhUkkbDGa6UhntJY=
 github.com/Xuanwo/templateutils v0.1.0/go.mod h1:OdE0DJ+CJxDBq6psX5DPV+gOZi8bhuHuVUpPCG++Wb8=
+<<<<<<< HEAD
```
Comment (xxchan) [inline go.sum]: please run `make build` again to clean it

**Comment by zxy-lgtm** [comment]
okok ,I will fix it soon.😃



---Original---
From: ***@***.***&gt;
Date: Tue, Sep 28, 2021 21:43 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [beyondstorage/go-service-onedrive] Basic Implement of OneDrive Storager (#17)




 
@xxchan commented on this pull request.
 
 
In go.sum:
 &gt;  github.com/Xuanwo/templateutils v0.1.0 h1:WpkWOqQtIQ2vAIpJLa727DdN8WtxhUkkbDGa6UhntJY=  github.com/Xuanwo/templateutils v0.1.0/go.mod h1:OdE0DJ+CJxDBq6psX5DPV+gOZi8bhuHuVUpPCG++Wb8= +<<<<<<< HEAD  
please run make build again to clean it
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by zxy-lgtm** [comment]
okok.



---Original---
From: ***@***.***&gt;
Date: Tue, Sep 28, 2021 11:27 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [beyondstorage/go-service-onedrive] Basic Implement of OneDrive Storager (#17)




 
I think we need to come up with a solution like feature flags or something. Can you just pin the go-integration-test to the older version and leave a comment on this comment?
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by xxchan** [comment]
Nice work! 

Next, let's add docs like: 
- https://github.com/beyondstorage/go-service-storj/pull/18 
- https://github.com/beyondstorage/site/pull/235
- https://github.com/beyondstorage/go-storage-example/pull/33

At the same time, don't forget the deadline of the final report of OSPP!
